app/demo-geant-integration/HitRootIO.cc

```diff
@@ -0,0 +1,208 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file demo-geant-integration/HitRootIO.cc
+//---------------------------------------------------------------------------//
+#include "HitRootIO.hh"
+
+#include <regex>
+#include <G4AutoDelete.hh>
+#include <G4Event.hh>
+#include <G4RunManager.hh>
+#include <G4Threading.hh>
+#include <TBranch.h>
+#include <TFile.h>
+#include <TObject.h>
+#include <TSystem.h>
+#include <TTree.h>
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/io/Logger.hh"
+#include "accel/ExceptionConverter.hh"
+#include "accel/SetupOptions.hh"
+
+#include "GlobalSetup.hh"
+#include "SensitiveHit.hh"
+
+namespace
+{
+// Mutex within the HitRootIO scope
+G4Mutex HitRootIOMutex = G4MUTEX_INITIALIZER;
+}  // namespace
+
+namespace demo_geant
+{
+G4ThreadLocal HitRootIO* HitRootIO::instance_ = nullptr;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create a ROOT output file for each worker except the master thread if MT
+ */
+HitRootIO::HitRootIO()
+{
+    G4AutoLock lock(&HitRootIOMutex);
+    gSystem->Load("libHitClassesDict");
+
+    file_name_ = std::regex_replace(
+        GlobalSetup::Instance()->GetSetupOptions()->output_file,
+        std::regex("json"),
+        "root");
+
+    if (G4Threading::IsWorkerThread())
+    {
+        file_name_ += std::to_string(G4Threading::G4GetThreadId());
+    }
+
+    if (G4Threading::IsWorkerThread()
+        || !G4Threading::IsMultithreadedApplication())
+    {
+        file_ = TFile::Open(file_name_.c_str(), "recreate");
+        CELER_VALIDATE(file_->IsOpen(), << "Failed to open " << file_name_);
+        tree_ = new TTree(
+            "Events", "Hit collections", this->SplitLevel(), file_);
+        G4AutoDelete::Register(tree_);
```

Comment:
What is the purpose of this registration? (I am not familiar which that interface.  When does it auto deletes)?  (Note the `_tree` will be delete by the TFile destructor (if still alive when it is run).

Reply:
G4AutoDelete::Register is guaranteed to delete thread private object when exiting thread.   May be no effect if tree_ deleted properly at the end of thread life time.

Reply:
That is okay to do if and only if you have are sure you called `Write` on the `tree_` before deleting it.   Similarly why not do the same for the file?

Reply:
Added G4AutoDelete for file_

Reply:
Although I originally wanted the "geant demo app" to use geant4 conventions, I'm now leaning toward using the best practices established in Celeritas for memory management etc. Would you be open to changing these to shared pointers? Is it also possible to construct the IO file using shared pointers in the `ActionInitialization` and/or detector initialization? Creating (and destroying through shared pointers) the objects rather than using a singleton and `AutoDelete` makes it easier to reason about the code, maintain it, and debug it.